boot/setefibootvars.go

```diff
@@ -0,0 +1,199 @@
+			continue
+		}
+		if varGUID != efi.GlobalVariable {
+			// Not an EFI global variable, so skip it
+			continue
+		}
+		varNumber, err := strconv.ParseUint(varName[4:], 16, 16)
+		if err != nil {
+			// Should not occur, since variable matched bootVarRegexp
+			return 0, err
+		}
+		// Since we never overwrite an existing variable, we can ignore
+		// variable attributes when reading the variable
+		varData, _, err := efiReadVariable(varName, varGUID)
+		if err != nil {
+			return 0, err
+		}
+		if bytes.Compare(optionData, varData) == 0 {
+			// existing variable already identical, use it
+			return uint16(varNumber), nil
+		}
+		usedBootNums[varNumber] = true
+	}
+	for bootNum := uint64(0); bootNum <= 0xFFFF; bootNum++ {
+		if usedBootNums[bootNum] == false {
+			return uint16(bootNum), ErrNoMatchingVariable
+		}
+	}
+	return 0, ErrAllBootNumsUsed
+}
+
+// Ensures that a Boot#### variable contains the given EFI load option.
+//
+// It may be the case than an existing boot variable already contains the
+// given load option, in which case that boot variable is reused. Otherwise,
+// finds the first unused boot variable number and uses it. Writes the load
+// option to that variable, and returns the number of the variable that was
+// used.
+func setEfiBootOptionVariable(loadOptionData []byte) (uint16, error) {
+	bootNum, err := findMatchingBootOption(loadOptionData)
+	if err == nil {
+		return bootNum, nil
+	} else if err != ErrNoMatchingVariable {
+		return 0, err
+	}
+	varName := fmt.Sprintf("Boot%04X", bootNum)
+	err = efiWriteVariable(varName, efi.GlobalVariable, defaultVarAttrs, loadOptionData)
+	return bootNum, err
+}
+
+// Reads the current BootOrder variable, inserts the given newBootNum at the
+// beginning of the number list (and removes it from later in the list if
+// it occurs) and writes the list as the new BootOrder variable.
+func SetEfiBootOrderVariable(newBootNum uint16) error {
+	origData, attrs, err := efiReadVariable("BootOrder", efi.GlobalVariable)
+	if err == efi.ErrVarNotExist {
+		attrs = defaultVarAttrs
+		origData = make([]byte, 0)
+	} else if err != nil {
+		return err
+	}
+	if len(origData)%2 != 0 {
+		return ErrInvalidBootOrder
+	}
+	var optionOffset int
+	for optionOffset = 0; optionOffset < len(origData); optionOffset += 2 {
+		bootNum := binary.LittleEndian.Uint16(origData[optionOffset:])
+		if newBootNum == bootNum {
+			break
+		}
+	}
+	var newData []byte
+	if optionOffset == 0 {
+		// newBootNum already at start, no need to re-write variable
+		return nil
+	} else if optionOffset == len(origData) {
+		// newBootNum not in original boot order
+		newData = make([]byte, len(origData)+2)
+		binary.LittleEndian.PutUint16(newData, newBootNum)
+		copy(newData[2:], origData)
+	} else {
+		newData = make([]byte, len(origData))
+		binary.LittleEndian.PutUint16(newData, newBootNum)
+		copy(newData[2:], origData[:optionOffset])
+		copy(newData[2+optionOffset:], origData[optionOffset+2:])
```

Comment:
For coherence, this should be `copy(newData[optionOffset+2:], origData[optionOffset+2:])`